internal/auth/client_kerberos.go

```diff
@@ -0,0 +1,366 @@
+		// TODO: Add option for strict acceptor check
+		ok, err := a.Verify(k.client.keytab, skew, hostAddr, nil)
+		if err != nil {
+			return nil, "", false, err
+		}
+		if !ok {
+			return nil, "", false, fmt.Errorf("Couldn't verify AP_REQ packet")
+		}
+
+		k.key = a.Authenticator.SubKey
+
+		ticket := a.Ticket
+		rep, err := krbmsg.NewAPRep(ticket, a.Authenticator)
+		if err != nil {
+			return nil, "", false, err
+		}
+
+		repToken := spnego.NewKRB5TokenAPREP()
+		repToken.APRep = rep
+		mar2, err:= repToken.Marshal()
+		asn1tools.AddASNAppTag(mar2, 060)
+		if err != nil {
+			return nil, "", false, err
+		}
+
+		if a.Authenticator.HasCredDelegation() {
+			var cred krbmsg.KRBCred
+			err = cred.Unmarshal(a.Authenticator.Delegation.Deleg)
+			if err != nil {
+				return nil, "", false, err
+			}
+			err = cred.DecryptEncPart(ticket.DecryptedEncPart.Key)
+			if err != nil {
+				return nil, "", false, err
+			}
+
+			cacheCreds, err := cred.ToCredentials()
+			if err != nil {
+				return nil, "", false, err
+			}
+			cache := credentials.CCacheFromCredentials(cacheCreds)
+			mar, err := cache.Marshal()
+			if err != nil {
+				return nil, "", false, err
+			}
+			k.credentials = mar
+		}
+
+		return mar2, k.username, false, nil
+	}
+
+	return nil, "", false, fmt.Errorf("Invalid token")
+}
+
+// RFC4462 Section 3.5
+type GSSAPIMicField struct {
+	SessionIdentifier string
+	Request byte
+	UserName string
+	Service string
+	Method string
+}
+
+func (mic *GSSAPIMicField) Unmarshal(b []byte) error {
+	err := ssh.Unmarshal(b, mic)
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *kerberosAuthContext) VerifyMIC(micField []byte, micToken []byte) error {
+	var t gssapi.MICToken
+	err := t.Unmarshal(micToken, false)
+	if err != nil {
+		return err
+	}
+	t.Payload = micField
+	verified, err := t.Verify(k.key, keyusage.GSSAPI_INITIATOR_SIGN)
+	if err != nil {
+		return err
+	}
+	if !verified {
+		return fmt.Errorf("Verify() returned unverified but no error")
+	}
+
+	// MIC is verified, but need to ensure usernames match
+	var field GSSAPIMicField
+	err = field.Unmarshal(micField)
+	if err != nil {
+		return err
+	}
+
+	// 50 = SSH_MSG_USERAUTH_REQUEST
+	if field.Request != 50 || field.Service != "ssh-connection" || field.Method != "gssapi-with-mic" {
+		return fmt.Errorf("Invalid MIC field")
+	}
+
+	if k.client.enforceUsername && field.UserName != k.username {
+		return fmt.Errorf("Invalid username for ticket")
+	}
+
+	k.success = true
+
+	return nil
+}
+
+func (k *kerberosAuthContext) DeleteSecContext() error {
+	return nil
+}
+
+func (k *kerberosAuthContext) AllowLogin(username string) error {
+	if !k.Success() {
+		return k.Error()
+	}
+
+	authzRequest := auth.AuthorizationRequest{
+		PrincipalUsername: k.username,
+		LoginUsername: username,
+		RemoteAddress: k.remoteAddr.String(),
+		ConnectionID: k.connectionId,
+	}
+	authzResponse := &auth.ResponseBody{}
+
+	statusCode, err := k.client.http.Post("/authz", authzRequest, authzResponse)
```

Comment:
This code should be in a retry and should be properly logged. See `processAuthWithRetry` for an example.